shtuff.py

```diff
@@ -21,15 +25,99 @@
     pass
 
 def main():
-    parser = argparse.ArgumentParser()
-    parser.add_argument('to_spawn', type=str, help='Command to spawn')
-    parser.add_argument('to_stuff', type=str, nargs='?', default='', help='Keys to stuff')
+    parser = argparse.ArgumentParser(
+        description="Shtuff will stuff commands into a shell à la tmux send-keys or screen stuff",
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        epilog=dedent("""\
+            Examples:
+                In shell A, run:
+                  $ shtuff as shell-a
+
+                In shell B, run:
+                  $ shtuff into shell-a "git status"
+
+                Observe how shell A ran "git status"
+
+                An example use case for `shtuff new` might be a setup script to
+                open a couple shells automatically. Consider this script:
+                  #!/usr/bin/env bash
+                  termite -e "shtuff new vim" &
+                  termite -e "shtuff new 'tail -f /var/log/somelog.log'" &
+
+                This script will open two terminals, one running vim, and one
+                running tail.
+        """),
+    )
     parser.add_argument('-v', '--version', action='version', version=__version__)
-    args = parser.parse_args()
-    spawn_and_stuff(args.to_spawn, args.to_stuff)
+    subparsers = parser.add_subparsers(metavar="action")
+
+    parser_as = subparsers.add_parser('as', help='become a receiving shell')
+    parser_as.add_argument('name', help='the current shell will take this name, use `shtuff into` to send commands here')
+    parser_as.set_defaults(func=shtuff_as)
+
+    def add_newline_argument(parser):
+        parser.add_argument('-n', dest="newline", action='store_false', help="don't send a trailing newline, similar to echo")
+
+    parser_into = subparsers.add_parser('into', help='send a command to a receiving shell')
+    parser_into.add_argument('name', help='the name of the shell to send the given command to')
+    parser_into.add_argument('cmd', help='the command to send to the shell')
+    add_newline_argument(parser_into)
+    parser_into.set_defaults(func=shtuff_into)
+
+    parser_new = subparsers.add_parser('new', help='start a new shell and immediately send a command')
+    parser_new.add_argument('cmd', help='start a new shell and immediately run the given command')
+    add_newline_argument(parser_new)
+    parser_new.set_defaults(func=shtuff_new)
+
+
+    args = vars(parser.parse_args())
+    func = args.pop('func')
+    func(**args)
```

Comment:
Historically, I like to keep parsing out of my main (in C and other languages). Here, it may appear silly, but I would make `main` be like this:

```python
def parse_args():
    ... all of the parsing code at the start of main...

def main():
    func, args = parse_args()
    return func(**args)
```

I added the return so your command line script can return a proper unix code.

Reply:
Yeah, we debated something like `parse_args()`. Probably good to just go ahead and do it. Good note about return, I didn't even think about that since it isn't my current use case, but since it's basically free, might as well.

Reply:
I don't see how adding a `return` statement here will help with exit status. `main` is just some function that we call at the bottom of this file, we don't do anything with it's return value.